File: sessionsync_script.go
package vaioidc

// The browser half of session sync, owned by the library (v0.21.0).
//
// # Why this is here and not in each consumer
//
// v0.20.0 shipped the server half and left the client half to consumers. The
// first consumer wrote 158 lines of JavaScript; eight more doors were queued to
// receive a copy of it. ⛔ That is the shape this fleet has been burned by
// repeatedly (one value, N hand-written producers), and it is at its worst here:
// the browser half is precisely the half with NO SERVER-SIDE SIGNAL. A copy that
// drifts, or one that is never updated when a verdict is added, fails in a
// browser console and nowhere else — on the product with the strictest CSP,
// which is the one most likely to need the fix.
//
// So the library serves the script, and a consumer's whole browser-side opt-in
// becomes one <script src> tag pointing at its own auth mount.
//
// # What a consumer still owns, and cannot be given
//
// ⚠️ THE CSP. `frame-src 'self' <AuthorizationOrigin()>` lives on the consumer's
// own page, produced by the consumer's own middleware, and no library can write
// it. That is why the beacon below exists: it is the one mechanism that turns
// "the frame was blocked" from a console message into a server log line.

import (
	_ "embed"
	"fmt"
	"log/slog"
	"net/http"
	"strings"
)

//go:embed sessionsync_script.js
var sessionSyncScriptSource string

// handleSessionSyncScript serves the client script, with both endpoint paths
// templated in from the mount this router was actually mounted at.
//
// ⭐ THE MOUNT IS DERIVED FROM THE REQUEST PATH, never configured and never
// assumed. The path the browser used to fetch this script is, by construction,
// the path that resolves to this router — so trimming the known suffix yields
// the consumer's real mount whatever prefix it chose. A configured mount would
// be a second producer of a value the router already knows, and the failure of
// getting it wrong is an iframe that loads a 404 and reports nothing.
//
// ⛔ THE DERIVED PREFIX IS VALIDATED BEFORE IT IS TEMPLATED INTO JAVASCRIPT.
// Everything reaching this point is a literal route match, so a hostile prefix
// is not reachable today — but "not reachable today" is the argument that makes
// the next refactor an injection. The charset refusal costs one function and
// removes the question.
//
// ⚠️ It assumes the browser-visible path equals the server-visible path. That
// holds on this fleet by decision (products are SUBDOMAINS, not paths, so no
// proxy rewrites a prefix away). A consumer behind a path-stripping proxy must
// not use this route.
func (a *Auth) handleSessionSyncScript(w http.ResponseWriter, r *http.Request) {
	mount, ok := mountFromScriptPath(r.URL.Path)
	if !ok {
		// Refusing is the only honest answer: a script whose endpoints were
		// guessed would load, run, and silently reconcile nothing.
		a.logger.Warn(logMsgSyncScriptPath,
			slog.String(logKeyComponent, logComponent),
			slog.String(logKeyPath, r.URL.Path),
		)
		http.Error(w, "session sync script: unusable mount path", http.StatusInternalServerError)
		return
	}

	body := strings.NewReplacer(
		scriptTokenSyncEndpoint, jsStringLiteral(mount+pathSessionSync),
		scriptTokenBlockedEndpoint, jsStringLiteral(mount+pathSessionSyncBlocked),
	).Replace(sessionSyncScriptSource)

	w.Header().Set(headerContentType, contentTypeJavaScript)
	// ⚠️ No-store rather than a long cache: the endpoints inside this body are
	// derived per request, and a cached copy of a body assembled for a different
	// mount is a script that reconciles against somebody else's routes. The file
	// is a few KB and is fetched once per page load.
	w.Header().Set(headerCacheControl, cacheControlNoStore)
	w.Header().Set(headerXContentTypeOptions, contentTypeOptionsNoSniff)
	w.WriteHeader(http.StatusOK)
	_, _ = w.Write([]byte(body))
}

// mountFromScriptPath recovers the consumer's mount prefix from the path the
// script was requested at, and refuses anything it cannot vouch for.
//
// Returns ("", false) when the path does not end in the script suffix, or when
// the recovered prefix contains a character that has no business in a route
// prefix. An empty prefix (mounted at root) is legitimate and returns ("", true)
// — ⛔ which is why the bool is separate from the string and must not be
// collapsed into an emptiness check.
func mountFromScriptPath(p string) (string, bool) {
	if !strings.HasSuffix(p, pathSessionSyncScript) {
		return "", false
	}
	mount := strings.TrimSuffix(p, pathSessionSyncScript)
	for _, c := range mount {
		switch {
		case c >= 'a' && c <= 'z', c >= 'A' && c <= 'Z', c >= '0' && c <= '9':
		case c == '/', c == '-', c == '_', c == '.':
		default:
			return "", false
		}
	}
	return mount, true
}

// jsStringLiteral renders s as a JavaScript string literal safe to substitute
// into a script body.
//
// ⚠️ It is belt AND braces: mountFromScriptPath has already refused every
// character this escapes. Both exist because the alternative is a comment
// asserting that the other one is enough.
func jsStringLiteral(s string) string {
	var b strings.Builder
	b.WriteByte('"')
	for _, r := range s {
		switch r {
		case '"', '\\', '\n', '\r', '<', '>', '&', '\u2028', '\u2029':
			// \uXXXX covers every case with one rule, including the two line
			// terminators that are newlines to a JavaScript parser and ordinary
			// characters to everything else.
			fmt.Fprintf(&b, "\\u%04X", r)
		default:
			b.WriteRune(r)
		}
	}
	b.WriteByte('"')
	return b.String()
}

// handleSessionSyncBlocked records that a browser gave up on reconciliation
// without ever reaching a verdict.
//
// ⭐ THIS IS THE ONLY SERVER-SIDE SIGNAL THE BROWSER HALF HAS. Every way session
// sync can fail in a browser — a missing `frame-src` directive on the parent
// page, a consumer middleware that overwrites the sync document's CSP, an
// `X-Frame-Options` header that forbids the frame, a provider that renders
// instead of redirecting — is reported to a browser console and NOWHERE ELSE.
// The product then keeps serving the previous user with every server-side signal
// green, which is exactly the failure this whole feature exists to remove.
//
// ⛔ IT REQUIRES A SESSION, and that is not politeness. An unauthenticated POST
// that writes a WARN line is a log-flooding primitive any page on the internet
// can aim at this origin. Requiring the session bounds it to people who are
// signed in here, which is also the only population whose report means anything:
// a blocked frame matters because a SIGNED-IN person is not being reconciled.
//
// ⚠️ It always answers 204, session or not. The browser can do nothing with a
// refusal, and an endpoint that discriminates in its status code tells an
// unauthenticated caller whether a given cookie is valid.
func (a *Auth) handleSessionSyncBlocked(w http.ResponseWriter, r *http.Request) {
	w.Header().Set(headerCacheControl, cacheControlNoStore)

	if _, err := readSessionCookie(r, a.sessionKey, a.cfg.CookieName); err != nil {
		w.WriteHeader(http.StatusNoContent)
		return
	}

	a.logger.Warn(logMsgSyncBlocked,
		slog.String(logKeyComponent, logComponent),
		slog.String(logKeyClientIP, clientIP(r)),
		slog.String(logKeyReason, reasonSyncBlocked),
	)
	w.WriteHeader(http.StatusNoContent)
}

File: sessionsync_script.js
/*
 * vaioidc session-sync client — the BROWSER half of identity reconciliation.
 *
 * ⛔ THIS FILE IS SERVED BY go-vai-oidc, NOT COPIED INTO CONSUMERS, AND THAT IS
 * THE POINT. Between v0.20.0 and v0.21.0 the only working implementation lived
 * as a 158-line hand-written copy in one consumer, about to be copied into eight
 * more. The browser half is precisely the half with no server-side signal: a
 * copy that drifts, or one that is never updated when a verdict is added, fails
 * in a browser console and nowhere else. One owner, reached by a pin bump.
 *
 * WHY IT EXISTS, and why it is not "poll harder":
 *
 * The server-side revalidation floor (Config.RevalidateInterval) asks the token
 * endpoint "is the session this cookie was minted from still alive?", using a
 * refresh token bound to THAT session. When somebody signs in as a DIFFERENT
 * person, Keycloak mints a NEW session and swaps the browser's identity cookie
 * WITHOUT terminating the old one — so the floor asks about the old session,
 * Keycloak truthfully answers "still alive", and the refreshed tokens come back
 * carrying the OLD person's subject. The product keeps serving the previous
 * user, correctly, indefinitely. No interval fixes that: the axis is the
 * QUESTION, not the frequency.
 *
 * Only a request that travels through the BROWSER carries the browser's Keycloak
 * cookie. This file makes that request, in a hidden same-origin iframe, and acts
 * on the verdict.
 *
 * ⛔ IT MUST NOT NAVIGATE TO LOGIN on a `switched` verdict. That would silently
 * adopt the new identity, which is exactly the surprise this reports instead of
 * performing (operator decision 2026-09-11: a tab must not change owner
 * underneath unsaved work). A reload renders the product's own signed-out state
 * and the person chooses.
 */
(function () {
  "use strict";

  // Both endpoints are templated in by the handler that serves this file, from
  // the mount the consumer actually chose. ⛔ Never transcribed by a consumer and
  // never guessed here: a script that assumes "/auth" is a script that is wrong
  // on every consumer that mounted anywhere else, silently.
  var ENDPOINT = __VAIOIDC_SYNC_ENDPOINT__;
  var BLOCKED_ENDPOINT = __VAIOIDC_BLOCKED_ENDPOINT__;

  var TIMER_MS = 5 * 60 * 1000;
  // A frame that never reports is the CSP-blocked case (frame-src missing, or a
  // provider that did not redirect). Reaping it bounds the damage to one hidden
  // iframe rather than one per focus event, forever.
  var TIMEOUT_MS = 20 * 1000;
  // After this many consecutive silent runs, stop and TELL THE SERVER. A
  // permanently blocked frame is a deployment defect that no amount of retrying
  // repairs, and retrying it on every focus is how a silent defect becomes a
  // visible leak.
  var MAX_SILENT = 3;

  var inFlight = false;
  var stopped = false;
  var silentRuns = 0;
  var timerID = null;

  function stop() {
    stopped = true;
    if (timerID !== null) {
      clearInterval(timerID);
      timerID = null;
    }
  }

  // ⭐ THE ONLY REPORT THIS FAILURE HAS. Everything above this line fails in a
  // browser console: a missing `frame-src` directive, a consumer middleware that
  // overwrites the sync document's CSP, a provider that renders instead of
  // redirecting. Each leaves the product serving the previous user with every
  // server-side signal green — the exact shape that cost this fleet two silent
  // outages. One beacon converts it into a server log line.
  function reportBlocked() {
    try {
      var body = JSON.stringify({ reason: "no_verdict", runs: silentRuns });
      if (navigator.sendBeacon) {
        navigator.sendBeacon(BLOCKED_ENDPOINT, new Blob([body], { type: "text/plain" }));
        return;
      }
      var xhr = new XMLHttpRequest();
      xhr.open("POST", BLOCKED_ENDPOINT, true);
      xhr.setRequestHeader("Content-Type", "text/plain");
      xhr.send(body);
    } catch (e) {
      /* the report is best-effort; never let it break the page */
    }
  }

  function act(result) {
    silentRuns = 0;
    // `disabled` is the server telling us this deployment did not opt in. It is
    // a truthful answer, not a failure — and re-asking it every five minutes
    // forever would be.
    if (result === "disabled") {
      stop();
      return;
    }
    // The session cookie is already gone server-side on both closing verdicts,
    // so a reload renders the signed-out page.
    if (result === "switched" || result === "signed_out") {
      window.location.reload();
    }
  }

  function sync() {
    if (stopped || inFlight || document.hidden) {
      return;
    }
    inFlight = true;

    var frame = document.createElement("iframe");
    frame.hidden = true;
    frame.setAttribute("aria-hidden", "true");
    frame.setAttribute("title", "session sync");
    frame.style.display = "none";

    var settled = false;
    var timer = null;

    function finish(result) {
      if (settled) {
        return;
      }
      settled = true;
      inFlight = false;
      if (timer !== null) {
        window.clearTimeout(timer);
      }
      window.removeEventListener("message", onMessage);
      if (frame.parentNode) {
        frame.parentNode.removeChild(frame);
      }
      if (result) {
        act(result);
        return;
      }
      silentRuns += 1;
      if (silentRuns >= MAX_SILENT) {
        stop();
        reportBlocked();
      }
    }

    function onMessage(event) {
      if (event.origin !== window.location.origin) {
        return;
      }
      var d = event.data;
      if (!d || d.source !== "vaioidc" || d.type !== "session-sync") {
        return;
      }
      finish(d.result);
    }

    window.addEventListener("message", onMessage);

    // Fallback for a consumer CSP that strips the document's inline script: the
    // verdict is also written to <html data-vaioidc-sync-result>, readable
    // because the final document is same-origin.
    frame.onload = function () {
      try {
        var el = frame.contentDocument && frame.contentDocument.documentElement;
        if (el && el.dataset && el.dataset.vaioidcSyncResult) {
          finish(el.dataset.vaioidcSyncResult);
        }
      } catch (e) {
        /* still on the provider's origin mid-flight — not an error */
      }
    };

    timer = window.setTimeout(function () {
      finish(null);
    }, TIMEOUT_MS);

    frame.src = ENDPOINT;
    document.body.appendChild(frame);
  }

  // On focus catches exactly the case this exists for: the person switched
  // accounts in another tab and came back to this one.
  window.addEventListener("focus", sync);
  document.addEventListener("visibilitychange", function () {
    if (!document.hidden) {
      sync();
    }
  });
  // The timer is an ADDITION, never the primary signal — browsers throttle
  // background timers, so it covers only a focused idle tab.
  timerID = window.setInterval(sync, TIMER_MS);

  // Exposed so an API client's 401 trap can reconcile before navigating to
  // login: a 401 caused by a switched identity should show the signed-out page,
  // not silently sign the browser's current person in.
  window.vaioidcSessionSync = sync;

  sync();
})();
